File: scala/src/main/scala/com/github/fidgetting/util/Jooq.scala
package com.github.fidgetting.util

import org.jooq.impl.DSL.trueCondition
import org.jooq.{Condition, Field, JSONB, Record, SelectConditionStep, SelectWhereStep}
import org.reactivestreams.Publisher
import reactor.core.publisher.{Flux, Mono}
import scalapb.{GeneratedMessage, GeneratedMessageCompanion}
import scalapb.json4s.JsonFormat

import scala.jdk.CollectionConverters.SeqHasAsJava

object Jooq {

  lazy val parser = JsonFormat.parser.ignoringUnknownFields
  lazy val printer = JsonFormat.printer

  extension (message: GeneratedMessage) {
    def toJSONB: JSONB =
      JSONB.valueOf(printer.print(message))
  }

  extension (jsonb: JSONB) {
    def toProto[T <: GeneratedMessage](implicit gmc: GeneratedMessageCompanion[T]): T =
      parser.fromJsonString(jsonb.data())
  }

  extension[R <: Record] (step: SelectWhereStep[R]) {
    def whereOpt[T](opt: Option[T])(f: T => Condition): SelectConditionStep[R] =
      step.where(opt.map(f).getOrElse(trueCondition()))
  }

  extension[T] (field: Field[T]) {
    def in(values: Iterable[_]): Condition = field.in(values.toSeq.asJava)
  }

  extension[T] (publisher: Publisher[T]) {
    def mono: Mono[T] = Mono.from(publisher)
    def flux: Flux[T] = Flux.from(publisher)
  }

}

File: java-reactive/src/main/java/com/github/fidgetting/util/Jooq.java
package com.github.fidgetting.util;

import com.google.protobuf.InvalidProtocolBufferException;
import com.google.protobuf.Message;
import com.google.protobuf.util.JsonFormat;
import org.jooq.JSONB;

import static com.github.fidgetting.util.GRPC.exception;
import static io.grpc.Status.INTERNAL;

public class Jooq {

  private static final JsonFormat.Parser parser = JsonFormat.parser().ignoringUnknownFields();
  private static final JsonFormat.Printer printer = JsonFormat.printer().omittingInsignificantWhitespace();

  public static <T extends Message.Builder> T parse(JSONB data, T builder) {
    try {
      parser.merge(data.data(), builder);
      return builder;
    } catch(InvalidProtocolBufferException e) {
      throw exception(INTERNAL, "Unable to parse protobuf: " + data.data(), e);
    }
  }

  public static JSONB toJSONB(Message message) {
    try {
      return JSONB.valueOf(printer.print(message));
    } catch(InvalidProtocolBufferException e) {
      throw exception(INTERNAL, "Unable to serialize protobuf: " + message, e);
    }
  }

}
